### ISSUE TITLE
[bug] Gemma 4 global-attention RoPE inv_freq has wrong shape vs HF proportional RoPE

### ISSUE THREAD

**Issue by jeje-soy-michu** [NVIDIA-NeMo/Megatron-Bridge#4195]
### Problem

Gemma 4's `full_attention` (global) layers use `rope_type="proportional"` in the HF
config. HF's `_compute_proportional_rope_parameters` builds `inv_freq` of length
`head_dim // 2` — `int(head_dim * partial_rotary_factor) // 2` real frequencies
followed by zero-padding — so `rotate_half` pairs head-dim lanes `(i, i + head_dim//2)`.

`Gemma4RotaryEmbedding` (`src/megatron/bridge/models/gemma/gemma4_provider.py`) instead
passes `rotary_percent=global_rotary_percent (0.25)` to the parent `RotaryEmbedding` and
builds `inv_freq` with only the 64 real frequencies and **no zero-pad**. The 64 frequency
*values* match HF bit-for-bit, so value-level checks pass — but the **shape (64 vs 256) is
wrong**, so once a forward pass hits `rotate_half` the head-dim lanes are paired
incorrectly, producing per-global-layer cosine drift that compounds into measurable
last-token logit divergence vs HF reference.

### Scope — affects all Gemma 4 variants, not just one

`Gemma4RotaryEmbedding` is instantiated in a single place (`Gemma4ModelProvider.provide()`),
and every variant routes through it:

- **MoE `gemma-4-26B-A4B`** 
- **Dense `gemma-4-31B`**
- **Text / VL**

All shipped variants set `global_rotary_percent = 0.25` (from the config's
`partial_rotary_factor`), so all hit the buggy partial path. Only the **global /
`full_attention` layers** are affected; local/sliding layers use a separate full-rotary
`rope_local` (`rope_type="default"`) and are correct.

### Minimal repro

```shell
"""Megatron Bridge Gemma4 global RoPE inv_freq shape != HF. Deps: torch, transformers>=5.6 (CPU-only)."""
import torch
from transformers.modeling_rope_utils import _compute_proportional_rope_parameters
from transformers.models.gemma4.configuration_gemma4 import Gemma4TextConfig

HEAD_DIM, THETA, PRF = 512, 1_000_000.0, 0.25  # global ("full_attention") params, shared by MoE/Dense/VL

cfg = Gemma4TextConfig()
cfg.head_dim, cfg.partial_rotary_factor = HEAD_DIM, PRF
cfg.rope_parameters = {
    "full_attention":    {"partial_rotary_factor": PRF, "rope_theta": THETA, "rope_type": "proportional"},
    "sliding_attention": {"rope_theta": 10000.0, "rope_type": "default"},
}
hf, _ = _compute_proportional_rope_parameters(cfg, device=torch.device("cpu"), layer_type="full_attention")

# Verbatim from gemma4_provider.py Gemma4RotaryEmbedding.__init__ on main:
dim = int(HEAD_DIM * PRF)  # 128
mg = 1.0 / (THETA ** (torch.arange(0, dim, 2, dtype=torch.float32) / HEAD_DIM))

print(f"HF   inv_freq: shape={tuple(hf.shape)}  nonzero={int((hf!=0).sum())}  zeros={int((hf==0).sum())}")
print(f"main inv_freq: shape={tuple(mg.shape)}")
print("shapes equal?        ", tuple(hf.shape) == tuple(mg.shape))
print("main == HF[:64] vals?", torch.allclose(mg, hf[:64], atol=1e-8, rtol=1e-6))
assert hf.shape[0] == HEAD_DIM // 2 and mg.shape[0] != hf.shape[0], "BUG"
print(f"=> BUG: HF expects length {HEAD_DIM//2}, main produces {mg.shape[0]}.")
```

Output:

```
HF   inv_freq: shape=(256,)  nonzero=64  zeros=192
main inv_freq: shape=(64,)
shapes equal?         False
main == HF[:64] vals? True
=> BUG: HF expects length 256, main produces 64.
```

### Expected behavior

Megatron's global `inv_freq` should match HF's proportional layout exactly: length
`head_dim//2` (256) = 64 real frequencies + 192 zeros, so `rotate_half` pairs
`(i, i+head_dim//2)` identically to HF — for every Gemma 4 variant.

### Affected area

area:model

### Regression?

No

### Environment

- Branch / commit: [`main @ 823b9519`](https://github.com/NVIDIA-NeMo/Megatron-Bridge/commit/823b9519)
- transformers: 5.6.0 (used for the repro) / repo pins 5.8.1
- GPU / machine: CPU (repro is CPU-only)

### Logs

```shell
HF   inv_freq: shape=(256,)  nonzero=64  zeros=192
main inv_freq: shape=(64,)
=> BUG: HF expects length 256, main produces 64.
```

**Comment by jeje-soy-michu** [comment]
No longer needed, the fix was already merged https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4148

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #4148 - feat(model): add Gemma-4 E4B support (layer spec, checkpoint loader, parity check) (merged 2026-06-17)
- **Issue**: #4195 - Gemma 4 global-attention RoPE inv_freq has wrong shape vs HF proportional RoPE

### 2. Root Cause
`Gemma4RotaryEmbedding` built `inv_freq` with only the real frequencies (length 64 via `rotary_percent=0.25`) and no zero-padding, while HF's proportional RoPE builds `inv_freq` of length `head_dim // 2` (real frequencies + zero-pad) so `rotate_half` pairs lanes `(i, i + head_dim//2)`. Values matched HF bit-for-bit but the shape (64 vs 256) mispaired head-dim lanes in `rotate_half`, producing per-global-layer cosine drift and last-token logit divergence vs HF. Resolved by the Gemma-4 E4B rework in #4148 (reporter confirmed: "the fix was already merged").

### 3. Trigger Configuration
Any Gemma-4 variant forward pass on `full_attention` (global) layers — all variants route through `Gemma4ModelProvider.provide()`.

### 4. Nature of the Bug
**Classification**: CODE BUG - numerics (RoPE shape/pairing), silent divergence class

### 5. Existing Test Coverage
**In Fix PR**: YES - #4148 added the Gemma4 parity-check example (HF vs Megatron logit comparison), the assertion class that catches RoPE mispairing.
**In NMFW Tests**: YES - `testcases/hermes/test_gemma4_parity_check_e4b.py` wraps that parity check (asserts max logit diff within atol vs HF reference — the exact corrupted quantity for this bug class).

### 6. Coverage Assessment
| Test | What it covers | Covers this bug? |
|-----------|--------|------------|
| `testcases/hermes/test_gemma4_parity_check_e4b.py` (NMFW) | HF↔Megatron forward-pass logit parity for Gemma-4 E4B | Yes — asserts the failing quantity |
| `testcases/hermes/test_gemma4_conversion_sh.py` (NMFW) | weight-mapping roundtrip | Partial (would not catch a forward-pass pairing error) |

Note: the NMFW parity test currently FAILS on latest main — it is detecting a residual Gemma4 E4B text-mode parity gap (max |logit diff| 1.64 > atol 1.0, run 20260705T225141Z) which we are tracking separately as a candidate product bug. That is the coverage working as intended: the parity assertion guards exactly this divergence class.

### 7. New Regression Test
Not needed - the parity-check test asserts the exact failing quantity for this bug class.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE
